src/main/scala/mesosphere/marathon/core/election/impl/ElectionServiceBase.scala

```diff
@@ -0,0 +1,258 @@
+}
+
+abstract class ElectionServiceBase(
+    config: MarathonConf,
+    system: ActorSystem,
+    eventStream: EventStream,
+    metrics: Metrics = new Metrics(new MetricRegistry),
+    electionCallbacks: Seq[ElectionCallback] = Seq.empty,
+    backoff: Backoff,
+    shutdownHooks: ShutdownHooks) extends ElectionService {
+  import ElectionServiceBase._
+
+  private lazy val log = LoggerFactory.getLogger(getClass.getName)
+
+  private[impl] var state: State = Idle(candidate = None)
+
+  import scala.concurrent.ExecutionContext.Implicits.global
+
+  def leaderHostPortImpl: Option[String]
+
+  val getLeaderDataTimer: Timer =
+    metrics.timer(metrics.name(MetricPrefixes.SERVICE, getClass, "current-leader-host-port"))
+
+  final override def leaderHostPort: Option[String] = getLeaderDataTimer {
+    synchronized {
+      try {
+        leaderHostPortImpl
+      }
+      catch {
+        case NonFatal(e) =>
+          log.error("error while getting current leader", e)
+          None
+      }
+    }
+  }
+
+  override def isLeader: Boolean = synchronized {
+    state match {
+      case Leading(_, _) => true
+      case _             => false
+    }
+  }
+
+  // scalastyle:off cyclomatic.complexity
+  override def abdicateLeadership(error: Boolean = false, reoffer: Boolean = false): Unit = synchronized {
+    state match {
+      case Leading(candidate, abdicate) =>
+        log.info(s"Abdicating leadership while leading (reoffer=$reoffer)")
+        state = Abdicating(candidate, reoffer)
+        abdicate(error)
+      case Abdicating(candidate, alreadyReoffering, candidateWasStarted) =>
+        log.info("Abdicating leadership while already in process of abdicating" +
+          s" (reoffer=${alreadyReoffering || reoffer})")
+        state = Abdicating(candidate, alreadyReoffering || reoffer, candidateWasStarted)
+      case Offering(candidate) =>
+        log.info(s"Canceling leadership offer waiting for backoff (reoffer=$reoffer)")
+        state = Abdicating(candidate, reoffer)
+      case Offered(candidate) =>
+        log.info(s"Abdicating leadership while candidating (reoffer=$reoffer)")
+        state = Abdicating(candidate, reoffer)
+      case Idle(candidate) =>
+        log.info(s"Abdicating leadership while being NO candidate (reoffer=$reoffer)")
+        if (reoffer) {
+          candidate match {
+            case None    => log.error("Cannot reoffer leadership without being a leadership candidate")
+            case Some(c) => offerLeadership(c)
+          }
+        }
+    }
+  }
+  // scalastyle:on
+
+  protected def offerLeadershipImpl(): Unit
+
+  private def setOfferState(offeringCase: => Unit, idleCase: => Unit): Unit = synchronized {
+    state match {
+      case Abdicating(candidate, reoffer, candidateWasStarted) =>
+        log.error("Will reoffer leadership after abdicating")
+        state = Abdicating(candidate, reoffer = true, candidateWasStarted)
+      case Leading(candidate, abdicate) =>
```

Comment:
I don't think we have a convention about this in our style guide, but you can use `_` if you won't use the values.